### PR TITLE
[Feat] Add support and testcase for HunyuanImage-3.0-Instruct-Distil

### PR DESCRIPTION
**Pull request by zengchuang-hw**
## Purpose

Add support for HunyuanImage-3.0-Distil (CFG distilled) and MeanFlow distilled models.
Params including "cfg_distilled" and "use_meanflow" added to hunyuan-image-3.0. 
Reference: https://github.com/Tencent-Hunyuan/HunyuanImage-3.0/blob/main/hunyuan_image_3/modeling_hunyuan_image_3.py

**Model changes:**
- Add `cfg_distilled` and `use_meanflow` config flags in `HunyuanImage3Config`
- Add `guidance_emb` and `timestep_r_emb` embedding modules for distilled models
- Implement `instantiate_guidance_tokens()` and `instantiate_timestep_r_tokens()` methods
- Skip negative CFG prefill for CFG distilled models (cfg_factor=1, no negative branch)
- Disable CFG parallel for distilled models
- Update tokenizer output to include `timesteps_r_scatter_index`

**Test changes:**
- Reorganize `THRESHOLDS` dict by model variant (`default`, `Instruct`, `quant`)
- Add `_get_thresholds()` helper to select appropriate thresholds based on model type
- Replace hardcoded "L20x Reference" values with thresholds dict in test output tables
- Consolidate quant test threshold constants into `THRESHOLDS["quant"]`
- Update reference image and CoT text files for distil model

## Test Plan
1. hunyuan UT test

2. DiT only accuracy test 

3. it2i e2e accuracy test 
```bash
# Run HunyuanImage accuracy test with distil model (default)
pytest -s -v tests/e2e/accuracy/test_hunyuan_image3.py 
```

## Test Result
✔1. Hunyuan UT test passed
```
collected 106 items                                                                                                                                                                                                                                                                                                      

test_hunyuan_fused_moe.py ......                                                                                                                                                                                                                                                                                   [  5%]
test_hunyuan_image3_distil_inputs.py ........                                                                                                                                                                                                                                                                      [ 13%]
test_hunyuan_image3_it2i_ar_format.py .                                                                                                                                                                                                                                                                            [ 14%]
test_hunyuan_image3_it2i_multi_image.py ...................................                                                                                                                                                                                                                                        [ 47%]
test_hunyuan_image3_sampler.py ................                                                                                                                                                                                                                                                                    [ 62%]
test_image_kv_cache_manager.py .......                                                                                                                                                                                                                                                                             [ 68%]
test_prompt_utils.py .................................        
=========================================================================================================================================== 106 passed, 22 warnings in 27.43s ============================================================================================================================================    
```

2. DiT only accuracy test.

✔3. E2E accuracy test passed with distilled model weights.
```
--- Running Summary
================================================================================================================================= 2 passed, 2 skipped, 18 warnings in 561.75s (0:09:21) ==================================================================================================================================
```
```
+-----------------------------+---------+-------------+
| Metric                      |   Value |   Threshold |
+=============================+=========+=============+
| COT similarity to reference |  0.9575 |        0.9  |
+-----------------------------+---------+-------------+
| COT prefix match            | 21      |       10    |
+-----------------------------+---------+-------------+
| Image-Image similarity      | 94.15   |       90    |
+-----------------------------+---------+-------------+
| SSIM                        |  0.2619 |        0.26 |
+-----------------------------+---------+-------------+
| PSNR (dB)                   | 13.48   |       12.5  |
+-----------------------------+---------+-------------+
PASSED

```

output image:
<img width="1286" height="724" alt="image" src="https://github.com/user-attachments/assets/bd4ffbed-3c2b-4d22-b24f-7aecb91bf3cb" />

*note that white border should be fixed by https://github.com/vllm-project/vllm-omni/pull/3626

baseline image:
<img width="1278" height="719" alt="image" src="https://github.com/user-attachments/assets/39e64354-fe15-4007-b478-4d52a860f7d2" />

note that baseline image comes form tencent/HunyuanImage-3.0 repo demo:
https://github.com/Tencent-Hunyuan/HunyuanImage-3.0/blob/main/run_demo_instruct_distil.sh
```
python3 run_image_gen.py \
    --model-id $MODEL_PATH  \
    --verbose 2 \
    --prompt "基于图一的logo，参考图二中冰箱贴的材质，制作一个新的冰箱贴" \
    --seed 42 \
    --reproduce \
    --bot-task think_recaption \
    --image-size "720x1280" \
    --use-system-prompt en_unified  \
    --image assets/demo_instruct_imgs/input_1_0.png,assets/demo_instruct_imgs/input_1_1.png \
    --save ./image_edit_1_distill.png \
    --moe-impl flashinfer \
    --diff-infer-steps 8 
```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan. Please provide the test scripts & test commands.
- [x] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>**

### CODE REVIEW

File: vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py [hunk 1]
```diff
@@ -3095,6 +3112,13 @@ def __call__(
                     tc_prev_mod = cur_mod.detach()
 
                 if should_compute:
+                    # Handle guidance for CFG distilled models
+                    if self.model.config.cfg_distilled:
+                        model_kwargs["guidance"] = torch.tensor(
+                            [1000.0 * self._guidance_scale],
+                            device=self.device,
+                            dtype=torch.bfloat16,
+                        ).repeat(latent_model_input.shape[0])
```
[the remark this answers is not in the document]
Comment (zengchuang-hw) [reply vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py:3178]: 👍 modified in model_inputs.update

**Comment by zengchuang-hw** [comment]
Though testcases not all passed , this problems are caused mainly by https://github.com/vllm-project/vllm-omni/issues/4138 and https://github.com/vllm-project/vllm-omni/pull/4134. this PR is ready for review. @Bounty-hunter PTAL

**Comment by zengchuang-hw** [comment]
@TaffyOfficial PTAL

**Comment by TaffyOfficial** [comment]
[P1] Step-execution path will break after the new num_special_tokens contract.
In pipeline_hunyuan_image3.py, _update_model_kwargs_for_generation() now requires model_kwargs["num_special_tokens"], but prepare_encode() still only sets num_image_tokens and never seeds num_special_tokens. First denoise step can reach the update call and then KeyError before step 2. It also keeps the old meaning of num_image_tokens by adding timestep/guidance tokens, while _generate() now splits image tokens and special tokens separately.
Relevant lines: pipeline_hunyuan_image3.py:1591-1599, pipeline_hunyuan_image3.py:1696-1702, pipeline_hunyuan_image3.py:2025-2030.

[P1] Distil/MeanFlow semantics are only wired into the normal generation loop, not step execution.
Normal loop sets timesteps_r per step and injects guidance for cfg_distilled, but _denoise_step_group() never does either before prepare_inputs_for_generation(). It also stores _STEP_CFG_FACTOR as 1 + int(guidance_scale > 1.0), so CFG-distilled requests still get duplicated rows and external CFG in step mode. That contradicts the PR’s own “cfg_factor=1 / no negative branch” behavior.
Relevant lines: pipeline_hunyuan_image3.py:2079-2085, pipeline_hunyuan_image3.py:2231-2291, hunyuan_image3_transformer.py:3165-3204.

[P1] AR KV reuse does not shift the newly added scatter indexes.
_truncate_reused_prefix() subtracts positive_reuse_len from gen_timestep_scatter_index, but leaves guidance_scatter_index and timesteps_r_scatter_index absolute. For distil/MeanFlow with AR KV reuse, first DiT forward will scatter into the wrong token position, or out of range, after input_ids, image_mask, and position_ids have already been truncated.
Relevant lines: hunyuan_image3_transformer.py:2906-2914.

**Review by Bounty-hunter — APPROVED** [review]
LGTM

**Review by Bounty-hunter — APPROVED** [review]

**Review by zengchuang-hw — COMMENTED** [review]
Reviewed head `440fbfa7f4a51c9680b2520868d83767ca79190d` (9 files, +569/-97). CI is green (3.11/3.12 builds, pre-commit, DCO); merge is blocked only on required review.

**P1/P2 - AR KV reuse breaks the new distilled scatter indices** (inline on pipeline_hunyuan_image3.py:1867): when the image stage reuses the AR-stage KV cache, `_truncate_reused_prefix` (hunyuan_image3_transformer.py:2907) rebases `gen_timestep_scatter_index` by `-positive_reuse_len` but leaves `guidance_scatter_index` / `timesteps_r_scatter_index` at absolute positions. Step 0 then scatters over the truncated sequence (wrong column or out-of-bounds), and the position-id rebuild in `_update_model_kwargs_for_generation` has the same problem. Reachable with the e2e test's `need_send_cache`/`need_recv_cache`: the distil CoT contains think/recaption, so `think_recaption_end_pos` is set and `positive_reuse_len > 0`.

**Notes (non-blocking):**
- `stop_token_ids: [128025]` was removed from the AR config globally, so the non-distill Instruct run is also affected (not only Distil). Please confirm Instruct e2e still terminates on `<eoi>`/EOS rather than being bounded by `max_tokens=8192`.
- `_get_thresholds()` always returns `THRESHOLDS["default"]`; the `quant`/`default` split is a dead abstraction - inline it.
- `get_timestep_r` and the `instantiate_*` helpers match the upstream Tencent pipeline; no issue found there.
- The new unit test only checks key pass-through on a mock (inline on test_hunyuan_image3_distil_inputs.py:18); it does not cover the scatter/KV-reuse path.

**Validation gap:** could not run the new unit test in this environment (`diffusers` not installed); eager `pytest tests/diffusion/models/hunyuan_image3/test_hunyuan_image3_distil_inputs.py` is still recommended.

File: vllm_omni/diffusion/models/hunyuan_image3/pipeline_hunyuan_image3.py [hunk 2]
```diff
@@ -1739,11 +1863,23 @@ def forward_call(
                     inputs_embeds, images, timestep, image_mask
                 )
                 inputs_embeds = self.instantiate_timestep_tokens(inputs_embeds, timestep, gen_timestep_scatter_index)
+                # Instantiate guidance and timestep_r tokens for distilled models
+                inputs_embeds = self.instantiate_guidance_tokens(inputs_embeds, guidance, guidance_scatter_index)
```
Comment (zengchuang-hw) [inline vllm_omni/diffusion/models/hunyuan_image3/pipeline_hunyuan_image3.py:1696]: AR KV reuse does not rebase the new scatter indices. `_truncate_reused_prefix` (hunyuan_image3_transformer.py:2907) shifts `gen_timestep_scatter_index` by `-positive_reuse_len`, but `guidance_scatter_index` and `timesteps_r_scatter_index` keep absolute positions. The diffusion loop truncates `input_ids` before step 0, so on `first_step` this call scatters the guidance embedding at an absolute column over the truncated `inputs_embeds` (wrong column, or out-of-bounds when `abs_pos >= truncated_seq_len`), and `_update_model_kwargs_for_generation` (pipeline_hunyuan_image3.py:1671-1677) has the same issue when rebuilding position ids. This is reachable with the e2e test's `need_send_cache`/`need_recv_cache`: the distil CoT contains think/recaption, so `think_recaption_end_pos` is set and `positive_reuse_len > 0`. Suggestion: subtract `positive_reuse_len` from both indices in `_truncate_reused_prefix` (mirroring line 2907) and add a KV-reuse + distilled unit test.

File: tests/diffusion/models/hunyuan_image3/test_hunyuan_image3_distil_inputs.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+pytestmark = [pytest.mark.core_model, pytest.mark.cpu]
+
+
+class MockHunyuanImage3Pipeline:
+    """Minimal mock that replicates prepare_inputs_for_generation method."""
+
+    def __init__(self):
+        # Mock config for distilled models
+        self.config = type("Config", (), {})()
+        self.hf_config = type("HFConfig", (), {"cfg_distilled": True, "use_meanflow": True})()
```
Comment (zengchuang-hw) [inline tests/diffusion/models/hunyuan_image3/test_hunyuan_image3_distil_inputs.py:22]: The test only asserts that distilled kwargs are passed through on a mock pipeline; it never exercises the real `instantiate_guidance_tokens` / `instantiate_timestep_r_tokens` scatter against actual embedding shapes, nor the AR KV-reuse truncation path where the new scatter indices must be rebased (see pipeline_hunyuan_image3.py:1867). Suggest adding at least one test with real embedding shapes and a truncated-prefix (KV reuse) scenario to cover the wrong-column/out-of-bounds case.

**Comment by TaffyOfficial** [comment]
@vllm-omni-review-bot

**Comment by hsliuustc0106** [comment]
Validation attempt on head `0e8a75b45f796c43ca4cb9c0cb7b6569ede86439`.

Command:
```bash
pytest -s -v tests/e2e/accuracy/test_hunyuan_image3.py
```

Environment: `z00586778`, Python 3.12.13, vLLM 0.27.0, 4× NVIDIA L20X exposed as AR TP=2 + DiT TP/EP=2 (`CUDA_VISIBLE_DEVICES=0,3,4,5`). The other GPUs on this shared host were occupied, so I could not run the intended 8-GPU topology safely.

Result (reproduced twice): `1 failed, 3 skipped`.
- `test_image_to_image_alignment_online`: server did not become ready. During DiT initialization both workers completed NCCL/EP setup and entered Hunyuan model construction, then received a shutdown signal; rank 0 exited with code 130 and the parent reported `Engine core initialization failed` / `EOFError`.
- Offline accuracy was skipped by its 8-card hardware marker; the two quant cases were skipped because the quant model env vars were not set.
- Retry summary: `1 failed, 3 skipped, 14 warnings in 758.76s (0:12:38)`.

No inference request or COT/CLIP/SSIM/PSNR assertion was reached, and there was no Python traceback or OOM from the model path before shutdown. This run therefore does **not** verify or disprove the PR's accuracy claim; it needs to be repeated on an exclusive 8×L20X/H100 allocation.

**Comment by zengchuang-hw** [comment]
## Validation update: 4 GPUs

Validated PR head `4195ed3d5c46e5a1ac4301b80ab654570a77fcac` after merging the latest `upstream/main` (`bbb3ae01a`).

Environment:
- vLLM `0.28.0`
- PyTorch `2.13.0+cu130`
- 4 GPUs split as AR TP=2 and DiT TP/EP=2

### Online accuracy

```bash
gpu run --gpus 4 -- pytest -vv -s tests/e2e/accuracy/test_hunyuan_image3.py
```

`test_image_to_image_alignment_online`: **PASSED**

- CoT semantic similarity: `0.9482` >= `0.9`
- Text prefix match: `21` >= `10`
- Image similarity: `94.8521` >= `90`
- SSIM: `0.2384` >= `0.23`
- PSNR: `13.13` >= `12.5`

Standard file summary: `1 passed, 3 skipped in 666.58s`.

### Offline accuracy on 4 GPUs

The offline case is still decorated with `num_cards=8`, so pytest skips it under a normal 4-GPU allocation. I temporarily changed only that local hardware marker to `num_cards=4`, ran the offline case, and restored the marker before committing.

```bash
gpu run --gpus 4 -- pytest -vv -s tests/e2e/accuracy/test_hunyuan_image3.py::test_image_to_image_alignment
```

`test_image_to_image_alignment`: **PASSED** (`1 passed in 487.78s`)

- CoT semantic similarity: `0.9580` >= `0.9`
- Text prefix match: `21` >= `10`
- Image similarity: `96.6263` >= `90`
- SSIM: `0.2697` >= `0.23`
- PSNR: `14.02` >= `12.5`

The two quantized accuracy cases were not run because `HUNYUAN_IMAGE3_RUN_QUANT_ACCURACY` and the required quantized model environment variables were not configured. The core BF16 online and offline paths both passed on 4 GPUs.

**Comment by zengchuang-hw** [comment]
@Bounty-hunter PTAL again. rebase newest version and update testcase

**Review by Bounty-hunter — APPROVED** [review]
LGTM

**Review by Gaohan123 — APPROVED** [review]
LGTM. Thanks

**Comment by zengchuang-hw** [comment]
CI Error has nothing to do witch this PR, update branch to rerun it again
